File: Java/homework-5/src/com/testts/Folder.java
package com.dimevagg.homework5;

import java.util.ArrayList;

public class Folder extends AbstractFileSystemNode {
    ArrayList<FileSystemNode> childrenList = new ArrayList<FileSystemNode>();
    public Folder(String name, FileSystemNode ... childNodes) {
        this.name = name;
        addChildren(childNodes);
    }

    public void addChildren(FileSystemNode ... childNodes) {
        for(int i=0; i < childNodes.length; i++) {
            FileSystemNode children = childNodes[i];
            children.setParent(this);
            childrenList.add(children);
        }
    }

    public ArrayList<FileSystemNode> getChildren() {
        return childrenList;
    }

    public File searchFile(String name) {
        // не самый лучший вариант поиска, но всё же
        File result = null;
        for(FileSystemNode node: this.getChildren()){
            if(node instanceof File) {
                if(node.getName() != name)
                    continue;
                result = (File)node;
                break;
            }
            Folder recursive = (Folder)node;
            result = recursive.searchFile(name);
            if(result != null)
                break;
        }
        return result;
    }
}

File: FileSystemTutorial/src/main/kotlin/Folder.kt
import java.util.ArrayList

class Folder(name: String, vararg childNodes: FileSystemNode) : AbstractFileSystemNode() {
    var childrenList = ArrayList<FileSystemNode>()

    fun addChildren(vararg childNodes: FileSystemNode) {
        for (i in 0 until childNodes.size) {
            val children = childNodes[i]
            children.setNodeParent(this)
            childrenList.add(children)
        }
    }

    fun searchFile(name: String): File? {
        // не самый лучший вариант поиска, но всё же
        var result: File? = null
        for (node in childrenList) {
            if (node is File) {
                if (node.getNodeName() !== name) continue
                result = node
                break
            }
            val recursive = node as Folder
            result = recursive.searchFile(name)
            if (result != null) break
        }
        return result
    }

    init {
        this.name = name
        addChildren(*childNodes)
    }
}